### PR TITLE
Close #LGVISIUM-54: Groundwater extraction

### PR DESCRIPTION
**Pull request by dcleres**
The following code allows you to extract the groundwater level information from the files in the Zurich dataset and add this information to the metadata of the borehole profile. 

The current performance is: 

```
'groundwater_depth_accuracy': 0.24375
'groundwater_depth_precision': 0.5571428571428572
'groundwater_depth_recall': 0.302325581395348
 'groundwater_depth_f1': 0.39195979899497485
'groundwater_depth_tp': 39
'groundwater_depth_fp': 31
'groundwater_depth_fn': 90
```

During the work on this issue, it was also detected that some files have wrong labels. Therefore, the performance might actually be better, than the metrics are suggesting. 

The chosen approach was to use basic keyword matching to identify the location of the groundwater information and then to parse the text around this field. 

The current limitations of this code are: 
- When the groundwater is displayed as a logo only without the keyword, the groundwater level cannot be assessed
- When the text extraction did not find the keyword or the groundwater level, then it is also likely that the extraction is wrong

### CODE REVIEW

File: src/stratigraphy/groundwater/utility.py
Comment (dcleres) [inline src/stratigraphy/groundwater/utility.py:12]: I am not super happy about this construct. I think it became like this every time I added a new rule.

Comment (stijnvermeeren-swisstopo) [reply src/stratigraphy/groundwater/utility.py:12]: How about something like

```
patterns = [
    r"(\d{2}\.\d{2}\.\d{4})", # some comment
    ...
]
for pattern in patterns:
    date_match = regex.search(pattern, text)
    if date_match:
        return date_match.group(1)
```

That would get rid of most of the code duplication and the excessively nested if-clauses.

File: src/stratigraphy/benchmark/score.py
Comment (stijnvermeeren-swisstopo) [inline src/stratigraphy/benchmark/score.py:246]: This code is now more or less duplicated three times in this method (coordinates, ground water, ground water depth). Would it be possible to create a new method that computers the metrics for any of these properties (taking the exact property as a parameter) and that can then be called three times?

File: src/stratigraphy/main.py
Comment (stijnvermeeren-swisstopo) [inline src/stratigraphy/main.py:260]: We need to agree on what we output in the logger and how we format it, because this is starting to become rather messy.

File: src/stratigraphy/util/draw.py
Comment (stijnvermeeren-swisstopo) [inline src/stratigraphy/util/draw.py:151]: Will need to be adjusted after https://github.com/swisstopo/swissgeol-boreholes-dataextraction/pull/73 is merged

File: src/stratigraphy/groundwater/groundwater_extraction.py
Comment (stijnvermeeren-swisstopo) [inline src/stratigraphy/groundwater/groundwater_extraction.py:27]: Why not already parse the extracted string and use a https://docs.python.org/3/library/datetime.html#datetime.date object here?

File: src/stratigraphy/groundwater/groundwater_extraction.py
Comment (stijnvermeeren-swisstopo) [inline src/stratigraphy/groundwater/groundwater_extraction.py:110]: Essentially this is an "equals" method for GroundwaterInformation objects, correct?

File: src/stratigraphy/groundwater/groundwater_extraction.py
Comment (stijnvermeeren-swisstopo) [inline src/stratigraphy/groundwater/groundwater_extraction.py:208]: "top three" and "matches[:5]" don't seem to fit together?
What's the motivation for only considering an limited number of matches anyway?

File: src/stratigraphy/groundwater/groundwater_extraction.py
Comment (stijnvermeeren-swisstopo) [inline src/stratigraphy/groundwater/groundwater_extraction.py:269]: Why do this, instead of returning all extracted groundwater information?
The Zürich ground truth does contain at most one entry per borehole profile, but that is a limitation of the ground truth, not something that we actually want to imitate when extracting ground water data. Here, we do want to extract ALL available ground water level measurements, if possible.

**Review by stijnvermeeren-swisstopo — CHANGES_REQUESTED** [review]
The computer vision (shape detection / template matching) are not yet part of this implementation, is that correct? Could you summarize (here and maybe also in the Jira ticket) the results of your experiments with those methods, and why you did not include them in this implementation yet?

I would not include the Jupyter notebooks in the repository. If you have anything that you think will be useful regularly, then please integrate it in the main code base. If you think that we should proceed with one of the computer vision approaches, then you could create a new ticket and a corresponding new Git branch for that, and already include the code in that branch.

File: src/stratigraphy/groundwater/utility.py
Comment (dcleres) [reply src/stratigraphy/groundwater/utility.py:12]: I like this idea. Thank you! I will implement that!

File: src/stratigraphy/main.py
Comment (dcleres) [reply src/stratigraphy/main.py:260]: I would say we can print the performance metrics.

File: src/stratigraphy/util/draw.py
Comment (dcleres) [reply src/stratigraphy/util/draw.py:151]: Yes, indeed.

File: src/stratigraphy/groundwater/groundwater_extraction.py
Comment (dcleres) [reply src/stratigraphy/groundwater/groundwater_extraction.py:110]: Yes

File: src/stratigraphy/groundwater/utility.py
Comment (stijnvermeeren-swisstopo) [reply src/stratigraphy/groundwater/utility.py:12]: Also, why not immediately match all cases at once, like `(\d{1,2}\.\d{1,2}\.\d{2,4})`?

Also, we need to make sure that we don't stop the rexeg in the middle of a number. Because the current implementation only extracts "9.1.20" from "9.1.2018".

File: src/stratigraphy/groundwater/groundwater_extraction.py
Comment (dcleres) [reply src/stratigraphy/groundwater/groundwater_extraction.py:208]: Thank you for catching this. I started with the three best matches but realized that the performance was better with five. I think here the logic of limiting the number of matches was to improve the speed (not a big performance boost, but still) as well as to avoid using any match that could reduce the performance of extraction. Therefore, using the five best seemed like a good compromise. As you can see, the list of groundwater keywords is quite long, as there are many different representations in the different documents. This is a bit tricky to handle and the more words we have the higher is the chance for a false positive.

File: src/stratigraphy/groundwater/groundwater_extraction.py
Comment (dcleres) [reply src/stratigraphy/groundwater/groundwater_extraction.py:269]: Ah I see. I was trying always to return the latest groundwater measurement that was performed or the most complete (elevation, depth, and date). I observed that by returning the "best" / most complete, the accuracy also improved since, in general, for a false positive keyword match, it would be hard to find elevation/depth/date around the keyword.

**Comment by dcleres** [comment]
One of the challenges about the CV segmentation of the groundwater logo here is also that it is not only about finding the groundwater logo but also, once it is found, mapping it to the proper depth from the other information available in the borehole profile.

File: src/stratigraphy/benchmark/ground_truth.py
Comment (stijnvermeeren-swisstopo) [inline src/stratigraphy/benchmark/ground_truth.py:35]: I'm not such a fan of including groundwater data with the "metadata". For me, the groundwater data belongs to the main geological content of the borehole profile, not to the metadata. Therefore, I would rather keep them separate in the ground truth as well.

Comment (dcleres) [reply src/stratigraphy/benchmark/ground_truth.py:35]: I totally agree. I refactored the code to change this in our presentation.

File: src/stratigraphy/groundwater/groundwater_extraction.py
Comment (dcleres) [reply src/stratigraphy/groundwater/groundwater_extraction.py]: Implemented in the latest version of the code.

File: src/stratigraphy/groundwater/groundwater_extraction.py
Comment (dcleres) [reply src/stratigraphy/groundwater/groundwater_extraction.py]: I chose to overwrite the __eq__ operator instead to make it clearer.

**Comment by dcleres** [comment]
@stijnvermeeren-swisstopo, I was able to address most of your comments and, as discussed this morning, left the comment where we select the N-best matches open for now. I will make a separate assessment of this later today. I do believe the code is ready for a second review round.

File: src/stratigraphy/util/draw.py
Comment (dcleres) [reply src/stratigraphy/util/draw.py]: Merged the code.

**Comment by stijnvermeeren-swisstopo** [comment]
@dcleres I implemented extracting more than one ground water measurement per borehole profile, as I think this is quite an essential requirement. It also has a major influence on how we compute the scores. It is now no longer sufficient to just distinguish between "correct" and "incorrect", but we need to count TP, FP and FN, as we already do with the layers. The required a significant amount of changes in the code.

The metrics for groundwater extract have deteriorated a little bit, but I feel like we can improve them again by some relatively simple rules, such as removing duplicates in the extracted measurements, and having some validation rules to get rid of entries that are not plausible. It also seems like the depth and elevation values are sometimes mixed up (e.g. the elevation value is extracted as "depth"). I did not have time to try implementing any of these this yet. Could you have a go at that?

The metrics for coordinate extracted also were affected a little. I'm not 100% sure why. Probably it is related to how we deal with missing values / None values in the computations. It needs more investigation.

Further things included in my commit:
- Groundwater was still mixed up with the metadata in the ground truth classes, this is now kept separate
- Rename most occurrences of "groundwater_information" to simple "groundwater", as the word "information" does not add much clarity and it makes the variable names unnecessarily long.
- Use the same ISO 8601 date format "YYYY-MM-DD" everywhere, instead of using different formats in different places.
- Try to more cleanly separate the actual groundwater data for evaluation (depth, elevation, date) from the location in the PDF (which we only have for the extracted data, but not for the ground truth)

What I would like to do still in this ticket:
- Remove duplicates and add some simple validation rules for the extracted ground water measurements, to reduce the number of false positives.
- Reevaluate the choice of only considering the N-best matches.

What I would like to create follow-ups for:
- Find a better way to associate the "correctness" from the evaluation with the individual extracted values, so that we can show (red/green) the correctness of each extracted groundwater measurement also in the visualisation.
- Have more shared code for the computations of TP, FP, FN, F1 and other metrics for the layers on the one hand and for the groundwater and coordinates data on the other hand.
- Make the naming more consistent. (For example, ground truth JSON uses "date" while the predictions.json uses "measurement_date".)

**Comment by dcleres** [comment]
@stijnvermeeren-swisstopo thank you for looking into the metrics and how to improve them regarding the scenario we are facing. 

In the latest commit, I addressed the following issues: 
- Removed the `_information` extension in all occurences, as indeed the extension does not add a lot of value and just makes it harder to read
- Added the accuracy to the metrics. I think this got lost in the different rebases. 
- A hardcoded threshold for the depth at 200m was added; as discussed, this should only be a temporary fix for the occasional confusion of the extractor between elevation and depth of the groundwater. I do believe that once we manage the merge of the elevation PR (#74), we should be able to build a logic that is able to identify these confusion cases. Otherwise, I was also thinking that we could take advantage of the depth_column information where the depths are also extracted and recorded. If the depth of the groundwater is larger than the maximum depth from the depth_column, there might be a good chance that there is a problem. Adding this threshold decreased the number of False positives 68 to 59.
- I also reevaluate the choice of only considering the N-best matches: There was no effect on the metrics. 

Currently the metrics are: 

```
{
   "F1":0.8790759488149966,
   "precision":0.8845491726684556,
   "recall":0.8736700405245328,
   "depth_interval_accuracy":0.0,
   "de_F1":0.8790759488149966,
   "de_precision":0.8845491726684556,
   "de_recall":0.8736700405245328,
   "de_depth_interval_accuracy":0.0,
   "coordinate_accuracy":0.592814371257485,
   "coordinate_precision":0.9339622641509434,
   "coordinate_recall":0.61875,
   "coordinate_f1":0.7443609022556391,
   "coordinates_tp":99,
   "coordinates_fp":7,
   "coordinates_fn":61,
   "groundwater_accuracy":0.013392857142857142,
   "groundwater_precision":0.03,
   "groundwater_recall":0.023622047244094488,
   "groundwater_f1":0.026431718061674006,
   "groundwater_tp":3,
   "groundwater_fp":97,
   "groundwater_fn":124,
   "groundwater_depth_accuracy":0.22043010752688172,
   "groundwater_depth_precision":0.41,
   "groundwater_depth_recall":0.3228346456692913,
   "groundwater_depth_f1":0.36123348017621143,
   "groundwater_depth_tp":41,
   "groundwater_depth_fp":59,
   "groundwater_depth_fn":86
}
```

**Comment by stijnvermeeren-swisstopo** [comment]
@dcleres 
I removed the n-best cutoff, because indeed, this does not seem to make a difference anymore.
I also improved the deduplication logic. Previously, the same line could still be extracted multiple times, if it was matched multiple times with different values for "fuzzy_counts". This improved the "groundwater depth F1" metric by another 3%.

Even without those changes, my numbers are better than yours ("groundwater depth F1" is now at 47.7%). Are you using the latest ground truth file from https://jira.swisstopo.ch/browse/LGVISIUM-61 ?

I feel like many of the remaining mistakes are actually due to incomplete data in the ground truth. E.g. several groundwater measurements are correctly extracted, but the ground truth only contains one measurement. Maybe we can spend some time on fixing those in the ground truth as well? We can use the existing open ticket https://jira.swisstopo.ch/browse/LGVISIUM-61 for that.

Could you create follow-up tickets for the remaining three open points from my previous comment?

**Comment by dcleres** [comment]
> I removed the n-best cutoff, because indeed, this does not seem to make a difference anymore.

@stijnvermeeren-swisstopo thank you for double-checking this

> Even without those changes, my numbers are better than yours ("groundwater depth F1" is now at 47.7%). Are you using the latest ground truth file from https://jira.swisstopo.ch/browse/LGVISIUM-61 ?

Actually, I was not. I only saw that you made some changes right now in the comments from the issue. Thank you for that! 

> I feel like many of the remaining mistakes are actually due to incomplete data in the ground truth. E.g. several groundwater measurements are correctly extracted, but the ground truth only contains one measurement. Maybe we can spend some time on fixing those in the ground truth as well? We can use the existing open ticket https://jira.swisstopo.ch/browse/LGVISIUM-61 for that.

I feel so too. Yes indeed let's clean the groundtruth asap

> Could you create follow-up tickets for the remaining three open points from my previous comment?

Ofc 

Thank you for the several reviews and I happy we now reached a stage where we can merge the code!

**Review by stijnvermeeren-swisstopo — APPROVED** [review]

**Pull request merged**
